src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1199 @@
+
+          CompactLatticeWeight weight_offset;
+          weight_offset.SetWeight(
+              LatticeWeight(0, -state_label_forward_prob_[arc_append.olabel]));
+          vector<CompactLatticeWeight> weights = {
+              arc_append_mod.weight, arc_chunk.weight, olat->Final(prev_final_state),
+              weight_offset};
+          BaseFloat v1 = 0, v2 = 0;
+          for (auto &i : weights) v1 += i.Weight().Value1();
+          for (auto &i : weights) v2 += i.Weight().Value2();
+          vector<int32> s;
+          for (auto &i : weights)
+            s.insert(s.end(), i.String().begin(), i.String().end());
+
+          arc_append_mod.weight = CompactLatticeWeight(LatticeWeight(v1, v2), s);
+          arc_append_mod.olabel = 0;
+          arc_append_mod.ilabel = 0;
+          aiter.SetValue(arc_append_mod);
+        } // otherwise, it has been pruned
+        state_label_forward_prob_.erase(arc_append.olabel);
+      }
+      KALDI_ASSERT(prev_final_state != -1); // at least one arc should be appended
+      // making all unmodified remaining arcs of final_arc_list_prev_ be connected to
+      // a dead state
+      olat->SetFinal(prev_final_state, CompactLatticeWeight::Zero());
+    } else
+      olat->SetStart(0);
+    KALDI_VLOG(2) << "Frame: " << NumFramesDecoded()
+                  << " states of chunk: " << clat.NumStates()
+                  << " states of the lattice: " << olat->NumStates();
+  } // TODO: check in the case the last frame is det twice
+
+  last_get_lattice_frame_ = NumFramesDecoded();
+  // step 4: re-determinize the final lattice
+  if (redeterminize) {
+    DeterminizeLatticePrunedOptions det_opts;
+    det_opts.delta = config_.det_opts.delta;
+    det_opts.max_mem = config_.det_opts.max_mem;
+    Lattice lat;
+    ConvertLattice(*olat, &lat);
+    Invert(&lat);
+    if (lat.Properties(fst::kTopSorted, true) == 0) {
+      if (!TopSort(&lat)) {
+        // Cannot topologically sort the lattice -- determinization will fail.
+        KALDI_ERR << "Topological sorting of state-level lattice failed (probably"
+                  << " your lexicon has empty words or your LM has epsilon cycles"
+                  << ").";
+      }
+    }
+    if (!DeterminizeLatticePruned(lat, config_.lattice_beam, olat, det_opts))
+      KALDI_WARN << "Determinization finished earlier than the beam";
+    Connect(olat); // Remove unreachable states... there might be
+    KALDI_VLOG(2) << "states of the lattice: " << olat->NumStates();
+  }
+
+  return (olat->NumStates() != 0);
+}
+
+// It is modified from LatticeFasterDecoderTpl::GetRawLattice()
+template <typename FST, typename Token>
+bool LatticeIncrementalDecoderTpl<FST, Token>::GetRawLattice(
+    Lattice *ofst, bool use_final_probs, int32 frame_begin, int32 frame_end,
+    bool create_initial_state, bool create_final_state) {
+  typedef LatticeArc Arc;
+  typedef Arc::StateId StateId;
+  typedef Arc::Weight Weight;
+  typedef Arc::Label Label;
+
```

Comment:
This code is super unstructured; it's not a good practice to write code like this.  Better to break it into functions, classes, etc., with clear interfaces, and document things clearly as you go..  This may be part of why you can't find the bug.

Reply:
This function is mostly the same to GetRawLattice(), I add explanation and comments in the code to describe the difference